app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,203 @@
 
-    if Sidekiq::Batch::Status.new(bid).total.zero?
-      on_complete(bid, callback_options)
-    else
-      info("DeleteObjectHierarchyWorker#batch_success_callback retry job with the hierarchy of workers: #{caller_worker_hierarchy}")
-      retry_job wait: 5.minutes
+      [
+        "Plain-#{ar_object_str}",
+        *ar_object.background_deletion.map { "Association-#{ar_object_str}:#{_1}" },
+      ]
     end
   end
 
-  def destroy_and_delete_associations
-    Array(object.background_deletion).each do |association_config|
-      reflection = BackgroundDeletion::Reflection.new(association_config)
-      next unless destroyable_association?(reflection.name)
+  # @param hierarchy [Array<String>] something like ["Plain-Service-1234", "Association-Service-1234:plans" ...]
+  # @note processes deleting a hierarchy of objects and reschedules itself at current progress after a time limit
+  def perform(*hierarchy)
+    return compatibility(*hierarchy) unless hierarchy.first.is_a?(String)
 
-      ReflectionDestroyer.new(object, reflection, caller_worker_hierarchy).destroy_later
+    started = now
+    while hierarchy.present? && now - started < WORK_TIME_LIMIT_SECONDS
+      Rails.logger.info "Starting background deletion iteration with: #{hierarchy.join(' ')}"
+      handle_one_hierarchy_entry!(hierarchy)
     end
+
+    @remaining_hierarchy = hierarchy
+    # can be like this instead of in `after_perform`, the benefit is that we don't use undocumented order of callbacks
+    # return unless hierarchy.present?
+    #
+    # # depending on
+    # unlock(resource: lock_key)
+    # self.class.perform_later(*hierarchy)
   end
 
-  def destroyable_association?(_association)
-    true
+  # we can just use job.arguments.first but for compatibility mode we want it not to lock
+  def lock_key
+    first_argument = arguments.first
+    first_argument.is_a?(String) ? first_argument : Random.uuid
   end
 
   private
 
-  def called_from_provider_hierarchy?
-    return unless (tenant_id = object.tenant_id)
-    caller_worker_hierarchy.include?("Hierarchy-Account-#{tenant_id}")
+  def hierarchy_entries_for(...)
+    self.class.send(:hierarchy_entries_for, ...)
   end
 
-  def callback_options
-    { 'object_global_id' => object.to_global_id, 'caller_worker_hierarchy' => caller_worker_hierarchy }
-  end
-
-  class ReflectionDestroyer
-
-    def initialize(main_object, reflection, caller_worker_hierarchy)
-      @main_object = main_object
-      @reflection = reflection
-      @caller_worker_hierarchy = caller_worker_hierarchy
-    end
-
-    def destroy_later
-      reflection.many? ? destroy_has_many_association : destroy_has_one_association
+  # handles a single hierarchy entry
+  # @return [Array<String>] hierarchy for a newly discovered object from association to delete or empty array otherwise
+  def handle_one_hierarchy_entry!(hierarchy)
+    entry = hierarchy.pop
+    case entry
+    when /Plain-([:\w]+)-(\d+)/
```

Comment:
You use a macro for the regular expression below `ASSOCIATION_RE`, why not here?

Reply:
Because when using as a constant, the macros don't take effect. I thought better keep `$1` and `$2` usage consistent across the two code paths. idk what is better.

Reply:
I don't understand your answer, why can you do this:

```ruby
ASSOCIATION_RE = /Association-(?<klass>[:\w]+)-(?<id>\d+):(?<association>\w+)/
case
when ASSOCIATION_RE
```

but not this?:

```ruby
PLAIN_RE = /Plain-([:\w]+)-(\d+)/
case
when PLAIN_RE
```

Reply:
ah, got what you mean. Possible. Just the other one is used twice, and this is not used anywhere else. You find value in having it as a constant?

Reply:
I'd do it just for consistence, but I'm OK if you don't do it.

Reply:
I also like @jlledom 's suggestion.

Reply:
done